File: ios-host-app/SimBridgeHost/Data/ApiClient.swift
// ApiClient.swift
// SimBridgeHost
//
// URLSession async/await REST client matching Android ApiClient.kt.
// All network calls use structured concurrency with a 15-second timeout.

import Foundation

final class ApiClient: Sendable {

    private let prefs: Prefs
    private let session: URLSession

    init(prefs: Prefs) {
        self.prefs = prefs
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 15
        config.timeoutIntervalForResource = 15
        self.session = URLSession(configuration: config)
    }

    // MARK: - Public API

    /// POST /auth/login -- returns JWT token.
    func login(serverUrl: String, username: String, password: String) async throws -> LoginResponse {
        let url = URL(string: "\(serverUrl.trimmingSuffix("/"))/auth/login")!
        let body = LoginRequest(username: username, password: password)
        return try await post(url: url, body: body, token: nil)
    }

    /// POST /devices -- register this device as a host.
    func registerDevice(name: String) async throws -> DeviceResponse {
        let baseUrl = prefs.serverUrl
        let url = URL(string: "\(baseUrl)/devices")!
        let body = DeviceRegisterRequest(name: name, type: "host")
        return try await post(url: url, body: body, token: prefs.token)
    }

    /// POST /pair -- pair with a client device using a pairing code.
    func pair(code: String) async throws -> PairResponse {
        let baseUrl = prefs.serverUrl
        let url = URL(string: "\(baseUrl)/pair")!
        let body = PairRequest(code: code)
        return try await post(url: url, body: body, token: prefs.token)
    }

    // MARK: - Private

    private func post<RequestBody: Encodable, ResponseBody: Decodable>(
        url: URL,
        body: RequestBody,
        token: String?
    ) async throws -> ResponseBody {
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json; charset=utf-8", forHTTPHeaderField: "Content-Type")

        if let token = token, !token.isEmpty {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }

        let encoder = JSONEncoder()
        request.httpBody = try encoder.encode(body)

        let (data, response) = try await session.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw ApiError.invalidResponse
        }

        guard (200...299).contains(httpResponse.statusCode) else {
            let responseBody = String(data: data, encoding: .utf8) ?? "no body"
            throw ApiError.httpError(code: httpResponse.statusCode, message: responseBody)
        }

        let decoder = JSONDecoder()
        return try decoder.decode(ResponseBody.self, from: data)
    }
}

// MARK: - Error Types

enum ApiError: LocalizedError {
    case invalidResponse
    case httpError(code: Int, message: String)

    var errorDescription: String? {
        switch self {
        case .invalidResponse:
            return "Invalid server response"
        case .httpError(let code, let message):
            return "HTTP \(code): \(message)"
        }
    }
}

// MARK: - String Extension

private extension String {
    func trimmingSuffix(_ suffix: String) -> String {
        if hasSuffix(suffix) {
            return String(dropLast(suffix.count))
        }
        return self
    }
}

File: host-app/app/src/main/java/com/simbridge/host/data/ApiClient.kt
package com.simbridge.host.data

import com.google.gson.Gson
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.IOException
import java.util.concurrent.TimeUnit

class ApiClient(private val prefs: Prefs) {

    private val gson = Gson()
    private val jsonType = "application/json; charset=utf-8".toMediaType()

    private val client = OkHttpClient.Builder()
        .connectTimeout(15, TimeUnit.SECONDS)
        .readTimeout(15, TimeUnit.SECONDS)
        .writeTimeout(15, TimeUnit.SECONDS)
        .build()

    /**
     * POST /auth/login — returns JWT token.
     */
    fun login(serverUrl: String, username: String, password: String): Result<LoginResponse> {
        val body = gson.toJson(LoginRequest(username, password))
            .toRequestBody(jsonType)
        val request = Request.Builder()
            .url("${serverUrl.trimEnd('/')}/auth/login")
            .post(body)
            .build()
        return execute(request, LoginResponse::class.java)
    }

    /**
     * POST /devices — register this device as a host.
     */
    fun registerDevice(name: String): Result<DeviceResponse> {
        val body = gson.toJson(DeviceRegisterRequest(name, "host"))
            .toRequestBody(jsonType)
        val request = Request.Builder()
            .url("${prefs.serverUrl}/devices")
            .header("Authorization", "Bearer ${prefs.token}")
            .post(body)
            .build()
        return execute(request, DeviceResponse::class.java)
    }

    /**
     * POST /pair — pair with a client device using a pairing code.
     */
    fun pair(code: String): Result<PairResponse> {
        val body = gson.toJson(PairRequest(code))
            .toRequestBody(jsonType)
        val request = Request.Builder()
            .url("${prefs.serverUrl}/pair")
            .header("Authorization", "Bearer ${prefs.token}")
            .post(body)
            .build()
        return execute(request, PairResponse::class.java)
    }

    private fun <T> execute(request: Request, clazz: Class<T>): Result<T> {
        return try {
            val response = client.newCall(request).execute()
            val responseBody = response.body?.string()
            if (response.isSuccessful && responseBody != null) {
                Result.success(gson.fromJson(responseBody, clazz))
            } else {
                Result.failure(IOException("HTTP ${response.code}: ${responseBody ?: "no body"}"))
            }
        } catch (e: Exception) {
            Result.failure(e)
        }
    }
}
